### DISCUSSION TITLE
How to build docker container from source?

### DISCUSSION THREAD

**Discussion by Ethanol6** [SamTV12345/PodFetch#945] [Q&A]
I wouldn't mind *attempting* to contribute but my experience with docker is just in running other peoples containers, never building my own. My initial attempts to run `docker build` have not worked. I have not seen anything in the documentation about building Podfetch specifically.

I've tried this so far:

`sudo docker build -t podfetch-test -f Dockerfile_cross_postgres .`

Which results in:

```
=> ERROR [amd64 2/2] COPY ./target/x86_64-unknown-linux-musl/release/podfetch /app/podfetch                 0.0s
------
 > [amd64 2/2] COPY ./target/x86_64-unknown-linux-musl/release/podfetch /app/podfetch:
------
Dockerfile_cross_postgres:12
--------------------
  10 |     FROM base as amd64
  11 |     COPY --from=cache /etc/ssl/certs/ca-certificates.crt /etc/ssl/certs/
  12 | >>> COPY ./target/x86_64-unknown-linux-musl/release/podfetch /app/podfetch
  13 |     
  14 |     FROM base as armv7
--------------------
```

**Comment by SamTV12345** [comment]
> I wouldn't mind _attempting_ to contribute but my experience with docker is just in running other peoples containers, never building my own. My initial attempts to run `docker build` have not worked. I have not seen anything in the documentation about building Podfetch specifically.
> 
> I've tried this so far:
> 
> `sudo docker build -t podfetch-test -f Dockerfile_cross_postgres .`
> 
> Which results in:
> 
> ```
> => ERROR [amd64 2/2] COPY ./target/x86_64-unknown-linux-musl/release/podfetch /app/podfetch                 0.0s
> ------
>  > [amd64 2/2] COPY ./target/x86_64-unknown-linux-musl/release/podfetch /app/podfetch:
> ------
> Dockerfile_cross_postgres:12
> --------------------
>   10 |     FROM base as amd64
>   11 |     COPY --from=cache /etc/ssl/certs/ca-certificates.crt /etc/ssl/certs/
>   12 | >>> COPY ./target/x86_64-unknown-linux-musl/release/podfetch /app/podfetch
>   13 |     
>   14 |     FROM base as armv7
> --------------------
> ```

The problem with Docker builds for Rust is that you have to have everything in your Docker container. For that I used blackdex/rust-musl images that rely on the musl c compiler instead of libgcc that links dynamically things.
Additionally Rust generates a platform dependant code. 

So if you want to run PodFetch on your  own machine. You need to run the following commands which will prepare your executables

docker run --rm  -v armv7prod:/root/.cargo -v "$(pwd)":/home/rust/src blackdex/rust-musl:x86_64-musl-stable cargo build --release --features postgresql
docker run --rm -v x86prod:/root/.cargo -v "$(pwd)":/home/rust/src blackdex/rust-musl:aarch64-musl-stable cargo build --release --features postgresql
docker run --rm -v arm64prod:/root/.cargo -v "$(pwd)":/home/rust/src blackdex/rust-musl:armv7-musleabihf-stable cargo build --release --features postgresql


Now you should have all 3 executables for all platforms. After that you just do :

docker buildx bake --set "*.tags=samuel19982/podfetch:latest" --set '*.dockerfile=Dockerfile_cross_postgres' 

Then you should have the Docker image of PodFetch for all available platforms

**Reply by Ethanol6** [reply]
Thanks a lot. One issue. There is no `Cargo.toml` file in my `/home/rust/src`. In fact there is no `/home/rust/src` directory at all. I tried using `~/git/PodFetch` so it could find the Cargo.toml file in the git repo but I get the same error, it still says `/home/rust/src` is missing.

If I make the directory `/home/rust/src` and copy the Cargo.toml file from the git repo into it then it works but I get compilation errors when `/home/rust/src/static/index.html` can't be found. So there is some issue with my directory structure here.

**Reply by SamTV12345** [reply]
Oh. You need to have PodFetch cloned on your machine (git clone https://github.com/SamTV12345/PodFetch.git) and then execute the docker run command there. But before that you need to create the index.html file in static. This can be done by doing pnpm run build in the ui directory and copying everything from the dist folder to the static folder.

**Reply by Ethanol6** [reply]
Yes, PodFetch is cloned already. This seems to have worked except for `docker run --rm -v armv7prod:/root/.cargo -v "$(pwd)":/home/rust/src blackdex/rust-musl:x86_64-musl-stable cargo build --release --features postgresql` returns:
>error: failed to remove file `/home/rust/src/target/x86_64-unknown-linux-musl/release/podfetch`

And when running the final command `docker buildx bake --set ".tags=samuel19982/podfetch:latest" --set '.dockerfile=Dockerfile_cross_postgres'` I get this:
```
[+] Building 0.0s (1/1) FINISHED                                 docker:default
 => [internal] load local bake definitions                                 0.0s
 => => reading docker-compose.yml 402B / 402B                              0.0s
 => => reading docker-bake.hcl 209B / 209B                                 0.0s
ERROR: could not find any target matching ''
```

Appreciate all the help!

**Reply by BubbaFatAss** [reply]
Just in case you didn't figure this out - you need to modify the command a little.  I ended up running:
`docker buildx bake --set *.tags=bubbafatass/podfetch:latest --set *.dockerfile=Dockerfile_cross_postgres`
You might also have to follow the instructions on https://docs.docker.com/build/building/multi-platform/ to enable the containderd image store.  And install QEmu via
`docker run --privileged --rm tonistiigi/binfmt --install all`
Then I was able to build multiplatform images correctly!
